torch/testing/_internal/common_methods_invocations.py

```diff
@@ -944,6 +944,41 @@ def __init__(self,
                                             **kwargs)
         self.ref = ref
 
+def sample_inputs_foreach(self, device, dtype, N):
+    if dtype in [torch.bfloat16, torch.bool, torch.float16]:
+        tensors = [torch.randn((N, N), device=device).to(dtype) for _ in range(N)]
+    elif dtype in torch.testing.get_all_int_dtypes():
+        tensors = [torch.randint(1, 100, (N, N), device=device, dtype=dtype) for _ in range(N)]
+    else:
+        tensors = [torch.randn((N, N), device=device, dtype=dtype) for _ in range(N)]
+
+    return tensors
+
+class ForeachUnaryFuncInfo(OpInfo):
+    """Early version of a specialized OpInfo for foreach unary functions"""
+    def __init__(self,
+                 name,
+                 method,
+                 inplace,
+                 ref,  # torch reference function
+                 dtypes=floating_and_complex_types(),
+                 dtypesIfCPU=all_types_and_complex(),
+                 dtypesIfCUDA=floating_and_complex_types_and(torch.half),
```

Comment:
Just wondering:
- Why are the dtypesIfCPU different from dtypesIfCUDA? Generally if something works on CPU I'd expect it to work on CUDA as well
- What does "dtypes" do? We've already specified what the dtypes for CPU and CUDA are

Reply:
> Just wondering:
> 
> * Why are the dtypesIfCPU different from dtypesIfCUDA? Generally if something works on CPU I'd expect it to work on CUDA as well

Unfortunately that's often not the case for half and bfloat16 suppor.

> * What does "dtypes" do? We've already specified what the dtypes for CPU and CUDA are

We also run meta tests and XLA (TPU) tests and support other hardware using our tests, and dtypes is for them.

Reply:
> Unfortunately that's often not the case for half and bfloat16 support

Those exceptions make sense. I'm wondering if there is a reason why dtypesIfCPU includes integral types by default but dtypesIfCUDA doesn't include those by default

Reply:
The problem with integral types and CUDA is that foreach doesn't allow type promotion right now (coming in next PRs of this stack). So with int input we cant return float result, therefore we have some restrictions that will be lifted later in this stack.